### PR TITLE
Xin 85 process qc tc

### PR DESCRIPTION
**Pull request by wjrjerome**

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 1]
```diff
@@ -157,21 +158,23 @@ func (x *XDPoS_v2) VoteHandler(voteMsg *utils.Vote) error {
...
-	thresholdReached, numberOfVotesInPool, hookError := x.votePool.Add(voteMsg)
-	if hookError != nil {
-		log.Error("Error while adding vote message to the pool, ", hookError)
-		return hookError
+	thresholdReached, numberOfVotesInPool, pooledVotes := x.votePool.Add(voteMsg)
+	if thresholdReached {
+		log.Debug("Vote pool threashold reached: %v, number of items in the pool: %v", thresholdReached, numberOfVotesInPool)
+		err := x.onVotePoolThresholdReached(chain, pooledVotes, voteMsg)
+		if err != nil {
+			return nil
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go:166]: return err

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 2]
```diff
@@ -307,22 +313,33 @@ func (x *XDPoS_v2) verifyTC(timeoutCert *utils.TimeoutCert) error {
...
+		x.highestQuorumCert = quorumCert
+	}
+	// 2. Get QC from header and update lockQuorumCert(lockQuorumCert is the parent of highestQC)
+	proposedBlockHeader := blockCahinReader.GetHeaderByHash(quorumCert.ProposedBlockInfo.Hash)
+	var decodedExtraField utils.ExtraFields_v2
+	err := utils.DecodeBytesExtraFields(proposedBlockHeader.Extra, &decodedExtraField)
+	if err != nil {
+		return err
+	}
+	x.lockQuorumCert = &decodedExtraField.QuorumCert
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go:330]:
```
if x.lockQuorumCert == nil || decodedExtraField.QuorumCert.ProposedBlockInfo.Round > x.lockQuorumCert.ProposedBlockInfo.Round {
    x.lockQuorumCert = &decodedExtraField.QuorumCert
}
```

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 3]
```diff
@@ -474,3 +491,31 @@ func (x *XDPoS_v2) getSyncInfo() utils.SyncInfo {
 		HighestTimeoutCert: x.highestTimeoutCert,
 	}
 }
+
+//TODO: find parent and grandparent and grandgrandparent block, check round number, if so, commit grandgrandparent
+func (x *XDPoS_v2) commitBlocks(blockCahinReader consensus.ChainReader, proposedBlockHeader *types.Header, proposedBlockRound *utils.Round) (bool, error) {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go:496]: blockCahinReader -> blockChainReader

Comment (wgr523) [reply consensus/XDPoS/engines/engine_v2/engine.go:496]: Round is int, so `proposedBlockRound utils.Round` without * is easier

File: eth/bfter/bft.go [hunk 4]
```diff
@@ -18,10 +19,11 @@ type broadcastTimeoutFn func(*utils.Timeout)
 type broadcastSyncInfoFn func(*utils.SyncInfo)
 
 type Bfter struct {
-	broadcastCh chan interface{}
-	quit        chan struct{}
-	consensus   ConsensusFns
-	broadcast   BroadcastFns
+	blockCahinReader consensus.ChainReader
```
Comment (wgr523) [inline eth/bfter/bft.go:22]: blockCahinReader -> blockChainReader

**Review by wgr523 — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 2 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go:330]: 👍

**Pull request merged**